python/tk_nuke/panels.py

```diff
@@ -0,0 +1,235 @@
+# 
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your 
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+Panel support for Nuke
+"""
+
+import os
+import sys
+import nuke
+import sgtk
+import nukescripts
+
+from sgtk.platform.qt import QtCore, QtGui
+
+class NukePanelWidget(nukescripts.panels.PythonPanel):
+    """
+    Wrapper class that sets up a panel widget in Nuke.
+    This panel widget wraps around a QT widget.
+    """
+    def __init__(self, bundle, dialog_name, panel_id, widget_class, *args, **kwargs):
+        """
+        Constructor.
+        
+        :param bundle: The app/engine/fraemwork that the dialog belongs to
+        :param dialog_name: Name to be displayed on the panel tab
+        :param panel_id: Unique id for this panel
+        :param widget_class: The class to be instantiated. Its constructor 
+                             should not take any parameters.
+        """
+        
+        # create a reference to the ToolkitWidgetWrapper class so that 
+        # we can refer to it safely using a single line of fully qualified
+        # python to return it:
+        # 
+        # __import__('sgtk')._panel_wrapper_class
+        #
+        # This necessary for the panel creation in Nuke
+        setattr(sgtk, "_panel_wrapper_class", ToolkitWidgetWrapper)
+        
+        # store various parameters on the sgtk object
+        # so that we can pass it to the constructor
+        # of the ToolkitWidgetWrapper() safely
+        setattr(sgtk, "_current_panel_class", widget_class)
+        setattr(sgtk, "_current_panel_id", panel_id)
+        setattr(sgtk, "_current_panel_args", args)
+        setattr(sgtk, "_current_panel_kwargs", kwargs)
+        setattr(sgtk, "_current_panel_bundle", bundle)
+
+        # Run parent constructor
+        nukescripts.panels.PythonPanel.__init__(self, dialog_name, panel_id)
+        
+        # now crate a one liner command that can safely refer to the widget class
+        cmd = "__import__('nukescripts').panels.WidgetKnob(__import__('sgtk')._panel_wrapper_class)"
+        
+        # and lastly tell nuke about our panel object 
+        self.customKnob = nuke.PyCustom_Knob(dialog_name, "", cmd)
+        self.addKnob(self.customKnob)
+
+
+class ToolkitWidgetWrapper(QtGui.QWidget):
+    """
+    Wrapper widget which wraps around a tk app widget
+    """
+    
+    def __init__(self):
+        """
+        Constructor
+        """
+        
+        # first, call the base class and let it do its thing.
+        QtGui.QWidget.__init__(self)
+        
+        # pick up members from the global sgtk namespace
+        # where it was set by the NukePanelWidget code above
+        panel_id = sgtk._current_panel_id
```

Comment:
I've seen this pattern in other places in Toolkit, but I never got to ask why that pattern is needed. A comment explaining why would be nice.